tls/extensions/s2n_client_psk.c

```diff
@@ -0,0 +1,250 @@
+
+    /* Calculating the binders requires a complete ClientHello, and at this point
+     * the extension size, extension list size, and message size are all blank.
+     *
+     * We'll write placeholder data to ensure the extension and extension list sizes
+     * are calculated correctly, then rewrite the binders with real data later. */
+    psk_params->binder_list_size = binder_list_size;
+    GUARD(s2n_stuffer_skip_write(out, binder_list_size));
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_match_psk_identity(struct s2n_array *known_psks, const struct s2n_blob *wire_identity,
+        struct s2n_psk **match)
+{
+    ENSURE_REF(match);
+    ENSURE_REF(wire_identity);
+    ENSURE_REF(known_psks);
+
+    *match = NULL;
+
+    for(size_t i = 0; i < known_psks->len; i++) {
+        struct s2n_psk *psk;
+        GUARD_RESULT(s2n_array_get(known_psks, i, (void**)&psk));
+        ENSURE_REF(psk);
+
+        if (wire_identity->size != psk->identity.size) {
+            continue;
+        }
+
+        if (memcmp(psk->identity.data, wire_identity->data, wire_identity->size) == 0) {
+            *match = psk;
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_client_psk_recv_identity_list(struct s2n_connection *conn, struct s2n_stuffer *wire_identities_in)
+{
+    ENSURE_REF(conn);
+
+    uint8_t wire_index = 0;
+    while(s2n_stuffer_data_available(wire_identities_in)) {
+        uint16_t identity_size;
+        GUARD_AS_RESULT(s2n_stuffer_read_uint16(wire_identities_in, &identity_size));
+
+        uint8_t *identity_data;
+        ENSURE_REF(identity_data = s2n_stuffer_raw_read(wire_identities_in, identity_size));
+
+        struct s2n_blob identity;
+        GUARD_AS_RESULT(s2n_blob_init(&identity, identity_data, identity_size));
+
+        /* TODO: Validate obfuscated_ticket_age when using session tickets.
+         *
+         * For identities established externally, an obfuscated_ticket_age of 0 SHOULD be
+         * used, and servers MUST ignore the value.
+         */
+        uint32_t obfuscated_ticket_age;
+        GUARD_AS_RESULT(s2n_stuffer_read_uint32(wire_identities_in, &obfuscated_ticket_age));
+
+        /* TODO: Implement the callback to choose a PSK.
+         *
+         * When we don't have a callback configured to choose a PSK, we should fall back to accepting
+         * the first PSK identity that also exists in our list of supported PSKs. */
+        GUARD_RESULT(s2n_match_psk_identity(&conn->psk_params.psk_list, &identity, &conn->psk_params.chosen_psk));
+
+        if (conn->psk_params.chosen_psk) {
+            conn->psk_params.chosen_psk_wire_index = wire_index;
+            return S2N_RESULT_OK;
+        }
+
+        wire_index++;
+    }
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_client_psk_recv_binder_list(struct s2n_connection *conn, struct s2n_blob *partial_client_hello,
+        struct s2n_stuffer *wire_binders_in)
+{
+    ENSURE_REF(conn);
```

Comment:
You must also add
```c
GUARD_AS_RESULT(s2n_blob_is_valid(partial_client_hello));
GUARD_AS_RESULT(s2n_stuffer_is_valid(wire_binders_in));
```
If `wire_binders_in` is NULL, you get a null-pointer dereference in line 159.

Reply:
We don't care if partial_client_hello is valid, or even NULL-- we just pass it on to s2n_psk_verify_binder, which is responsible for validating its own arguments.

Reply:
What about `wire_binders_in`? I'd say we need a valid stuffer for `wire_binders_in` in order to `s2n_stuffer_data_available` work properly.